cdap-securestore-spi/src/main/java/co/cask/cdap/securestore/spi/SecretsMetadataStore.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ *  use this file except in compliance with the License. You may obtain a copy of
+ *  the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *  License for the specific language governing permissions and limitations under
+ *  the License.
+ */
+
+package co.cask.cdap.securestore.spi;
+
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * Stores and retrieves metadata for the secrets.
+ *
+ * @param <T> the type of the metadata to be stored.
+ */
+public interface SecretsMetadataStore<T> {
+
+  /**
+   * Retrieves stored metadata for the secret.
+   *
+   * @param namespace the namespace to which secret belongs to
+   * @param name the name of the secret to be retrieved
+   * @return stored metadata
+   * @throws SecretNotFoundException if secret is not found in provided namespace
+   * @throws Exception if unable to retrieve the secret
+   */
+  T get(String namespace, String name) throws Exception;
+
+  /**
+   * Provides list of all the secrets for which metadata is stored.
+   *
+   * @param namespace the namespace to which secrets belong to
+   * @return list of all the secrets stored in provided namespace
+   * @throws Exception if not able to get metadata for the secrets in the provided namespace
+   */
+  Collection<T> list(String namespace) throws Exception;
+
+  /**
+   * Stores metadata for the secrets.
+   *
+   * @param namespace the namespace to which secret belong to
+   * @param name the name of the secret
+   * @param metadata the metadata to be stored
+   * @throws SecretAlreadyExistsException if the metadata for the secret is already stored
```

Comment:
So I cannot overwrite an existing data?

Reply:
Current implementation of SecureStore apis does not support updates: https://github.com/cdapio/cdap/blob/develop/cdap-api/src/main/java/co/cask/cdap/api/security/store/SecureStoreManager.java#L39

Reply:
It does not because the Hadoop KMS does not. However, for the SPI implementation who uses this API, why this storage class not supporting update? I don't why they are related.

Reply:
And for the Hadoop KMS implementation, it won't use this class, right?

Reply:
I kept it because properties for secret such as description, creationtime, user provided properties are stored in metadata too. If the implementation updates any of those values, it may change that information as well. Do you think we should still support updates in store @chtyim ?